core/src/main/java/ai/timefold/solver/core/impl/domain/valuerange/cache/ValueRangeCacheStrategy.java

```diff
@@ -25,7 +25,11 @@ public sealed interface ValueRangeCacheStrategy<Value_> permits HashSetValueRang
      * by combining the current cache instance with the one passed as an argument.
      */
     default void merge(@NonNull CountableValueRange<Value_> otherCache) {
-        otherCache.createOriginalIterator().forEachRemaining(this::add);
+        otherCache.createOriginalIterator().forEachRemaining(value -> {
+            if (value != null) {
+                this.add(value);
+            }
+        });
```

Comment:
I wonder why the value even is `null` - null is being added by a wrapping range. The wrapping range should be the top-level range, null should not show up here.

Reply:
The `otherCache` could be a `NullAllowingCountableValueRange` and return a null value.

Reply:
I understand that. But I wonder... why would we even cache null? Why is that code there? Maybe we can now move the null-wrapping code to a higher level of the resolver. No longer wrap the range, but have the logic whether null is allowed inside the resolver.

Reply:
The class `AllEntitiesListValueRange` merges all value ranges to generate a unique list of elements. To merge the ranges, we call `FromEntityPropertyValueRangeDescriptor#extractValueRange(...)`. This may return `NullAllowingCountableValueRange` if the planning list permits unassigned values. So, if we don't filter null values, they will be stored in the range cache.

Reply:
The fact that something has been done in some way doesn't mean it has to stay that way forever. :-)

`NullAllowing...` can be removed. `ValueRangeResolver` can be modified to include null when it knows that it should. It doesn't need to be part of any of the collections - it would simply be an additional "artifical" value. Just like it is in the `NullAllowing...Range`.